lib/src/main/java/graphql/nadel/validation/NadelPartitionValidation.kt

```diff
@@ -0,0 +1,126 @@
+                    parentObject,
+                    overallSchema
+                )
+            }
+
+        val partitionAppliedToSubscriptionField =
+            conditionalError(PartitionAppliedToSubscriptionField(parent, overallField)) {
+                parentObject.name == "Subscription"
+            }
+
+        val unsupportedOutputType = conditionalError(
+            PartitionAppliedToFieldWithUnsupportedOutputType(
+                parent,
+                overallField
+            )
+        ) {
+            !overallField.type.unwrapNonNull().isList && !isMutationPayloadType(overallField.type)
+        }
+
+        val invalidPartitionArgument = conditionalError(
+            InvalidPartitionArgument(
+                parent,
+                overallField
+            )
+        ) {
+            val pathToPartitionArg = overallField.getAppliedDirective(NadelDirectives.partitionDirectiveDefinition.name)
+                ?.getArgument("pathToSplitPoint")
+                ?.getValue<List<String>>()
+                ?: return@conditionalError true
+
+            if (pathToPartitionArg.isEmpty()) {
+                return@conditionalError true
+            }
+
+            val argumentRoot = overallField.getArgument(pathToPartitionArg[0])
+                ?: return@conditionalError true
+
+            var currentType = argumentRoot.type.unwrapNonNull()
+
+            // start at 1 because we've already checked the first item
+            for (i in 1 until pathToPartitionArg.size) {
+                val key = pathToPartitionArg[i]
+
+                val inputType = currentType as? GraphQLInputObjectType
+                    ?: return@conditionalError true
+
+                currentType = inputType.getField(key).type.unwrapNonNull()
+            }
+
+            !currentType.isList
+        }
+
+        return partitionAppliedOnUnsupportedField + partitionAppliedToSubscriptionField + unsupportedOutputType + invalidPartitionArgument
+    }
+
+    private fun isMutationPayloadType(type: GraphQLOutputType): Boolean {
+        val objectType = type.unwrapNonNull() as? GraphQLObjectType ?: return false
+
+        val containsSuccessField = objectType.fieldDefinitions
+            .filter { it.name == "success" }
+            .map { it.type.unwrapNonNull() }
+            .filterIsInstance<GraphQLScalarType>()
+            .filter { it.name == "Boolean" }
+            .size == 1
+
+        val allExtraFieldsAreOfTypeList = objectType.fieldDefinitions
+            .filter { it.name != "success" }
+            .all { it.type.unwrapNonNull().isList }
+
+        return containsSuccessField && allExtraFieldsAreOfTypeList
+    }
+
+    private fun conditionalError(
+        error: NadelSchemaValidationError,
+        conditionSupplier: () -> Boolean,
+    ): List<NadelSchemaValidationError> {
+        return if (conditionSupplier()) {
+            listOf(error)
```

Comment:
You could also make this a supplier and mark the function as `inline` for effectively zero cost lambda